Editor/Viewer while in a call (Viewer Overlay Call View)

ShGKme opened the pull request:
### ☑️ Resolves

* Fix #9295
* Fix #8551

### 🖼️ Screenshots

![image](https://user-images.githubusercontent.com/25978914/235152082-dda555c7-804a-4da2-b4e3-e751ede238cf.png)

### 🚧 Tasks

- [x] Create embedded view.
  - [x] Show auto-pilot view in embedded container
  - [x] Add local audio/video controls on embedded instead of participant's ones
  - [x] Add  draft for maximize button
- [x] Integrate embedded view with file view
  - [x] Switch to embedded view on file view open
  - [x] Close embedded view on file view close
  - [x] Close file view on maximize button 
- [x] Improvement
  - [x] Add minimize/maximize button to embedded view
  - [x] Add local view preview on embedded view
- [x] Design review
- [ ] Fix `ETA: unknown`
  - [ ] Fullscreen mode
  - [ ] With local screensharing
  - [ ] Receiving screensharing
- [x] Everything is fine - refactoring `ETA: 1 day`
  - [x] Move to a separated component
  - [x] Fix tests
  - [x] Make local media controls more reusable
  - [x] Clean-up

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included or not required
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

ShGKme commented:
## Ready for the first design review

![image_2023-04-27_23-03-21](https://user-images.githubusercontent.com/25978914/235002721-cb135e75-d4c6-401a-b2d8-22860024cbf4.png)

### The differences from the initial issue:

### 1. Buttons styles.
In the original feature request they have transparent background. But in most of similar places in 
both Talk and UI components we have buttons with no background. Background is visible on the hover.

### 2. Mic button has volume indicator same as in the top bar.

### Additional

- Click on the video works as maximize button
- Currently the view is a bit large, 300x400px. I'd say, we can make it resizable on a next step

## Current drawbacks

### 1. Limited call management

To make embedded view simple, there is no other management buttons. A user can switch to a main call view if needed. This is about:
  - Start / stop screensharing
  - Disable/enable other participants
  - General conversation management

### 2. No dynamic aspect ratio - fixed at 4:3

Embedded video works with autopilot. The video source may change. To make it static, aspect ratio is fixed, the video may be cut.

Also no dynamic aspect ratio for a video of yourself. It is very small anyway. Prefer to have a static layout here.

## Next steps

- Check edge cases
- Clean-up

nimishavijay commented:
Looks really good! Only a few points apart from the ones you mentioned:

- [ ] All the buttons need a hover feedback. This can be the same as the hover feedback for the mute/video buttons when you are in a talk call
- [ ] Creating new file from the chat opens the file without showing the viewer
- [ ] If the view is minimised, the maximise button is not visible anymore because the color is white on a white background. The color on this button should be `--color-main-text` (and shouood adjust automatically from dark mode as well)

![image](https://user-images.githubusercontent.com/52440189/235171020-e273d3f6-10af-48dd-aa7b-cd4f98921c87.png)


- [ ] If the sidebar is closed, the viewer should move to the side

![image](https://user-images.githubusercontent.com/52440189/235170437-fd859ed6-236b-4f22-8001-cdc9ba0b6664.png)

- [ ] on smaller screens the viewer is pretty big and covers almost half the screen, it can be reduced in width and height by 30%
 
![image](https://user-images.githubusercontent.com/52440189/235169322-a0a85d35-988e-4e97-943d-dc1ae187ded5.png)

It's very nice! :) Super cool!

ShGKme commented:
> All the buttons need a hover feedback. This can be the same as the hover feedback for the mute/video buttons when you are in a talk call

> If the view is minimised, the maximise button is not visible anymore because the color is white on a white background. The color on this button should be --color-main-text (and shouood adjust automatically from dark mode as well)

Fixed

![Light](https://user-images.githubusercontent.com/25978914/235258682-5e44c911-46e2-4b26-af70-dda7fe0938df.gif)

![Dark](https://user-images.githubusercontent.com/25978914/235258662-03624397-6e8e-488b-86dd-65514f498ec5.gif)

> Creating new file from the chat opens the file without showing the viewer

Good catch! Fixed.

> If the sidebar is closed, the viewer should move to the side

Fixed. I was sure, I checked it 😅

> on smaller screens the viewer is pretty big and covers almost half the screen, it can be reduced in width and height by 30%

Now it has floating with a 20% screen width between 250px and 400px:
- 250px on ~1280px screen and smaller
- 400px on ~1920px screen and larger

![image](https://user-images.githubusercontent.com/25978914/235243052-6d46204d-c4b1-4969-87d1-1cdbed8e33ff.png)

Antreesy reviewed (approved):
Super! Tested on web client.

Couple of remarks:
`Expand` button could inherit the same style as `Scroll to bottom` in chat, to be more visible
![image](https://user-images.githubusercontent.com/93392545/235865944-c15c4205-d946-46c4-86c0-7c71e1f1b504.png)
Round corners for local video are too big. Could we shrink it a bit (selector rule is on screenshot)?
![image](https://user-images.githubusercontent.com/93392545/235866469-09c69b1f-f8d9-4e48-a89d-0a3a40ce8500.png)

ShGKme commented on src/components/CallView/shared/ViewerOverlayCallView.vue:
```diff
@@ -0,0 +1,413 @@
+  -
+  - @author Grigorii Shartsev <me@shgk.me>
+  -
+  - @license GNU AGPL version 3 or any later version
```
I followed this: https://github.com/nextcloud/server/blob/master/contribute/HowToApplyALicense.md

cc @nickvergessen

nickvergessen replied:
Should be the machine readable one. If you try it in js files eslint will complain. But the eslint rule seems to skip .vue files.

We should also update the [HowToApplyALicense.md](https://github.com/nextcloud/server/blob/master/contribute/HowToApplyALicense.md) then.

ShGKme replied:
```
$ grep -lr "AGPL-3.0-or-later" src | wc -l
> 114

$ grep -lr "GNU AGPL version 3 or any later version" src | wc -l
> 130
```

Should we update all 130 files?

cc @nickvergessen

nimishavijay commented:
> Should we keep the button, or remove it, or make it looks like a close button?

I would keep it as you have implemented it. Currently the picture-in-picture view that exists has the same button in the same position. It might be confusing if suddenly it doesn't show up anymore, in case someone has been using this.

![image](https://user-images.githubusercontent.com/52440189/235898469-2ed941e9-44fc-47f8-9259-22ab2952b3c4.png)

What do you think? @jancborchardt 

> Expand button could inherit the same style as Scroll to bottom in chat, to be more visible

Totally agreed here!

> Round corners for local video are too big. Could we shrink it a bit 

New screenshot looks great, great catch!

Additionally would suggest we could also move the local video a few pixels down so that it is bottom aligned with the bottom of the audio/video buttons, and a few pixels to the right so that the margin in the bottom and right is equal.

![image](https://user-images.githubusercontent.com/52440189/235946483-c7b659e2-03fc-413a-b7ee-f70ef8d5653e.png)

ShGKme commented:
> Additionally would suggest we could also move the local video a few pixels down so that it is bottom aligned with the bottom of the audio/video buttons, and a few pixels to the right so that the margin in the bottom and right is equal.

Good catch, I also don't like then something is "almost but not quite".

Fixed

Small | Large
---|---
![image](https://user-images.githubusercontent.com/25978914/235948733-3b3e1937-8b65-4914-b8b6-f348cf5d93fb.png) | ![image](https://user-images.githubusercontent.com/25978914/235949029-459c48c5-62b0-4574-a76e-64c997da4e81.png)

nimishavijay commented:
Nice! And also move it a few px to the right so that the bottom margin and right margin are equal. currently there is slightly more space to the right than in the bottom :)

Antreesy commented:
> Nice! And also move it a few px to the right so that the bottom margin and right margin are equal. currently there is slightly more space to the right than in the bottom :)

Something like calc(`overlay border-radius`px - `localVideo border-radius`px) for 'bottom' and 'right' properties, so the frame around will be with constant thickness

ShGKme commented:
Sorry for a force push, it's was a 1 line typo in the last refactoring commit

nickvergessen reviewed (commented):
Otherwise solid work!

juliushaertl commented:
Nice, just tested this with text and collabora to see if there is anything we need to do from the office team side.

One issue I noticed though was that the modal does not play well with other elements that may still be visible from talk:
- Modals like the device settings are hidden behind the viewer modal
- Action menus are cut off as they are layered below the viewer modal

![Screenshot 2023-05-04 at 09 57 36](https://user-images.githubusercontent.com/3404133/236145278-190e06bd-7f2c-4394-879c-dad6fc67eaeb.png)

It might be an option to render viewer to a different root element as in https://github.com/nextcloud/server/pull/33347/files#diff-22f78918676c8422a8c807a774435aae85eb2f9feed73d4af629f194092d349bR157

nickvergessen commented:
That is still tracked in #8709 for a follow up

rriemann commented:
> IMO, it is fine. When the viewer is open, it is kind of a primary app on the page. You can return to the call view to setup anything and then return back to the file.

I recorded a video for #8709 that demonstrates the issue. Please take extra care for usability. People may not understand easily the concept of primary app and secondary apps and models.


https://user-images.githubusercontent.com/111932/231517426-bdb88451-519a-4090-9ceb-be0498ea6c1f.mp4

nickvergessen commented:
> I recorded a video for https://github.com/nextcloud/spreed/issues/8709 that demonstrates the issue. Please take extra care for usability. People may not understand easily the concept of primary app and secondary apps and models.

We are totally aware, that's also why #8709 is still open. But at least you can disable your mic/camera now when you joined prior to editing. So yeah, doesn't fix everything. But at at least one of the two cases.
